### DISCUSSION TITLE
Special Tablet Configuration

### DISCUSSION THREAD

**Discussion by ImPurpl3** [ppy/osu#29245] [Ideas] [answered]
## Background
- I use a Wacom CTH-460 with flipped USB endpoint
- For most CTH-460 tablets, the endpoint is not flipped, but there are many tablets with this weird bug
- Flipped endpoint is something I don't totally understand, but whenever I use OpenTabletDriver, I have to use a special configuration json for my tablet that makes the tablet digitizer information readable by OTD. 
- The configuration file **is not** configuring buttons to keybinds or for plugins and filters
- The configuration file essentially lets OTD know how to get the digitizer information and button states
- Many people in the OTD Discord have been having issues with flipped endpoint on their CTH-460s
- TL;DR some CTH-460s are weird and need special configuration for OpenTabletDriver in order work

## The Issue
- osu!lazer's built-in tablet support uses the default `CTH-460.json` configuration file
- This means that any CTH-460s with flipped endpoint are found by the game, but none of the digitizer information is read
- TL;DR some CTH-460 tablets are useless for osu!lazer

## The Proposed Solution
- If a CTH-460 is detected in osu!, a option appears with the option to change configurations for flipped endpoint
- With the option selected, the alternate CTH-460 configuration is loaded and the tablet works! :D
- A warning appears saying something like "This option is only used for the CTH-460 tablet with a flipped endpoint"
- See mockup (ooh fancy)

![mockup](https://github.com/user-attachments/assets/00970be4-9248-455b-ac28-237493b78083)

## Notes
- I'm not sure if this is the right place to discuss this issue as I feel like this is more of an OTD problem, BUT as osu!lazer doesn't allow for user created configurations, I figured this would be the place to propose a solution to this problem
- I read in the OTD Discord that their tablet config handler is too limited to detect flipped endpoint, so handling the issue with a configuration is easier
- Thank you for considering this idea! I know this is probably a pretty small issue, considering most people aren't using these rare flipped endpoint tablets; so if you have issues with this, I'd love to hear your feedback so I know what to do moving forward!

### My testing experience/procedure
- I am on Windows 10, lazer ver. 2024.731.0
- I did test osu!lazer's tablet support with a CTH-490 and it worked (confirmed tablet support works on my system)
- Tried putting my "Configurations" folder in different game folders to see if the configuration detection from OTD would use my config
   - This is a feature in OTD where you can create a folder named `Configurations` in the same directory as the main executable and it will read user created configs from there
- Tooled around with `input.json` but figured it isn't tied to the driver configurations
- Compared the flipped endpoint json with the original json to make sure I wasn't going crazy
- Tried compliling my own build of lazer, but had trouble getting the prerequisite resources necessary to develop due to my system being incredibly low on storage _(this is more of a me issue, but then I resorted to what I know best: using dnSpy)_
- Patched my working `CTH-460.json` into the `OpenTabletDriver.Configurations.dll` and had amazing results from it
   - As the game updates, I have to repatch, which is something I really don't mind doing because it doesn't take long
   - But if this idea gets implemented, it will save me some time, and could really help others with the same problem
- If I am needed for testing, I would love to help!
   - I'm fairly free this summer
   - I am getting more storage soon

### My Motivation
- I prefer to use the built-in support opposed to using OTD because I found that the input feels faster and more responsive
- For playing on osu!stable, I would use OTD with my config and raw input on because raw input eliminated a latency issue I was having, but this caused a conflict with my mouse sensitivity
   - I use windows with high mouse dpi, low pointer speed in windows settings
      - With raw input on, my mouse speed was very high
   - This means that I couldn't lower my in-game sensitivity to make mouse speed lower without also affecting my tablet sensitivity
   - _Mouse and tablet sensitivity tied together in the stable game settings :(_
- With lazer's tablet support, I can do raw input+low mouse sensitivity in-game and not worry about my tablet sensitivity
   - _Mouse and tablet sensitivity separated :D_
- I am a huge sucker for seeing numbers do silly things as I make progress, so when I found out that lazer has daily challenges, I really wanted to try lazer

### CTH-460.json Configuration File for Flipped Endpoint
- Sourced from the OpenTabletDriver Discord
[CTH-460.json](https://github.com/user-attachments/files/16464628/CTH-460.json)

**Comment by Stoppedpuma** [comment, marked as answer]
Please report this at the [OpenTabletDriver git repo](https://github.com/OpenTabletDriver/OpenTabletDriver) as this is an issue with OpenTabletDriver, not osu!.
